pkgs/applications/networking/mullvad/mullvad.nix

```diff
@@ -0,0 +1,95 @@
+{ lib
+, stdenv
+, writeText
+, rustPlatform
+, fetchFromGitHub
+, pkg-config
+, protobuf
+, makeWrapper
+, dbus
+, libnftnl
+, libmnl
+, libwg
+}:
+let
+  # result of running address_cache as of 02 Mar 2022
+  bootstrap-address-cache = writeText "api-ip-address.txt" ''
+    193.138.218.78:443
+    193.138.218.71:444
+    185.65.134.66:444
+    185.65.135.117:444
+    217.138.254.130:444
+    91.90.44.10:444
+  '';
+in
+rustPlatform.buildRustPackage rec {
+  pname = "mullvad";
+  version = "2022.1";
+
+  src = fetchFromGitHub {
+    owner = "mullvad";
+    repo = "mullvadvpn-app";
+    rev = version;
+    hash = "sha256-bLwuM3Qy2iStbXIvDEWp31vuiihSQThOej297XKo5Xc=";
+  };
+
+  cargoHash = "sha256-CBbm8cJHTjyvvzCFQfKmsE5d9N7azEm8nI6KeWLVaa8=";
+
+  nativeBuildInputs = [
+    pkg-config
+    protobuf
+    makeWrapper
+  ];
+
+  buildInputs = [
+    dbus.dev
+    libnftnl
+    libmnl
+  ];
+
+  # talpid-core wants libwg.a in build/lib/{triple}
+  preBuild = ''
+    dest=build/lib/${stdenv.targetPlatform.config}
+    mkdir -p $dest
+    ln -s ${libwg}/lib/libwg.a $dest
+  '';
+
+  postFixup =
+    # Place all binaries in the 'mullvad-' namespace, even though these
+    # specific binaries aren't used in the lifetime of the program.
+    # `address_cache` is used to generate the `api-ip-address.txt` file, which
+    # contains list of Mullvad API servers -- though we provide a "backup" of
+    # the output of this command, it could change at any time, so we want
+    # users to be able to regenerate the list at any time. (The daemon will
+    # refuse to start without this file.)
+    ''
+      for bin in address_cache relay_list translations-converter; do
+        mv "$out/bin/$bin" "$out/bin/mullvad-$bin"
+      done
+    '' +
+    # Put distributed assets in-place -- specifically, the
+    # bootstrap-address-cache is necessary; otherwise, the user will have to run
+    # the `address_cache` binary and move the contents into place at
+    # /var/cache/mullvad-vpn/api-ip-address.txt manually. It is unknown at this
+    # time whether the `ca.crt` is also necessary.
```

Comment:
I am pretty sure this is required because it is used for verifications

https://github.com/mullvad/mullvadvpn-app/search?q=ca.crt

https://github.com/mullvad/mullvadvpn-app/blob/c3834e02312bd0ad8ad03e3993dbaf046dfde94f/README.md#repository-structure

> ca.crt - The Mullvad relay server root CA. Bundled with the app and only OpenVPN relays signed by this CA are trusted

Reply:
OpenVPN doesn't even work with this... It seems like they require using their fork of OpenVPN... Sighhhhhhh.

https://github.com/mullvad/openvpn/blob/mullvad-patches

Trying to use OpenVPN with this fails: 

```
2022-03-23 20:23:15 PLUGIN_INIT: plugin /home/vin/workspace/vcs/nixpkgs/master/test-dir/libtalpid_openvpn_plugin.so expressed interest in unsupported plugin types: [want=0x00009005, have=0x00007fff]
```

where `0x00009005` == `0b1001000000000101`, which has the `1 << 15`th bit set (which is invalid in "normal" OpenVPN: https://github.com/OpenVPN/openvpn/blob/2612125d7cf5e3c8687a3fab8fba61670ac12f35/include/openvpn-plugin.h.in#L131, but implemented in their patchset: https://github.com/mullvad/openvpn/blob/b45b090c81e7b4f2dc938642af7a1e12f699f5c5/include/openvpn-plugin.h.in#L134).

...I really don't want to package their patchset for OpenVPN...

Reply:
> https://github.com/mullvad/openvpn/blob/mullvad-patches

The right link is https://github.com/mullvad/openvpn/tree/mullvad-patches

I think we only need those 7 patches https://github.com/OpenVPN/openvpn/compare/release/2.5...mullvad:mullvad-patches and we could override our openvpn.

Reply:
Please take a look at https://github.com/cole-h/nixpkgs/pull/3

Reply:
Thanks a ton for that help :heart: With that packaged, I'll work on making it actually work :P Should be easier now that the fork is packaged...